### PR TITLE
issue #25705 - extend doc type

### PR DESCRIPTION
**Pull request by gilmoskowitz**
issue #25705:
This set of changes replaces the hand-coded `select` in the `docinfo` view for reporting document associations. The new version calls a function to dynamically build and execute a query based on a new `doctype` table. This should allow extensions to add arbitrary document types by changing data rather than [core C++](https://github.com/xtuple/qt-client/blob/4_9_x/widgets/documents.cpp).

There are accompanying pull requests in xtuple/qt-client#666 and xtuple/private-extensions#595

### CODE REVIEW

File: foundation-database/public/patches/populate_source.sql [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                     'join itemsite on coitem_itemsite_id=itemsite_id join item on itemsite_item_id=item_id');
+select createDoctype(33, 'SHP',  'SHP', 'SHP', 'Ship To', 'shipto',  'shipto_id',  'shipto_num',         'cust_name',     'shipto_name',
+                     '', 'join custinfo on shipto_cust_id=cust_id', 'shipto_id', 'shipTo');
+select createDoctype(34, 'TE',   'TE', 'TE', 'Time Expense', 'tehead',  'tehead_id',  'tehead_number',      'emp_number', 'formatDate(tehead_weekending)',
+                     '', 'join emp on tehead_emp_id = emp_id', 'tehead_id', 'timeExpenseSheet');
+select createDoctype(35, 'TD', 'TODO', 'TODO', 'To-Do',               'todoitem','todoitem_id','todoitem_id::text',  'todoitem_name', 'firstline(todoitem_description)',
+                     '', '',     'todoitem_id', 'todoItem', 'MaintainPersonalToDoItems MaintainAllToDoItems');
+select createDoctype(36, 'TO',   'TO', 'TO', 'Transfer Order',      'tohead',  'tohead_id',  'tohead_number',      's.warehous_code', 'd.warehous_code',
+                     '', 'join whsinfo s on tohead_src_warehous_id = s.warehous_id join whsinfo d on tohead_dest.warehous_id = d.warehous_id',
```
Comment (bendiy) [inline foundation-database/public/patches/populate_source.sql]: `tohead_dest_warehous_id`, not `tohead_dest.warehous_id`?

File: foundation-database/public/patches/populate_source.sql [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+select createDoctype(24, 'J',    'J', 'J', 'Project',          'prj',     'prj_id',     'prj_number',          'prj_name',      'firstline(prj_descrip)',
+                     'core', '',     'prj_id',     'project',     'MaintainPersonalProjects MaintainAllProjects');
+select createDoctype(25, 'P',    'P', 'P', 'Purchase Order',   'pohead',  'pohead_id',  'pohead_number',       'vend_name',     'firstline(pohead_comments)',
+                     'core', 'join vendinfo on pohead_vend_id = vend_id', 'pohead_id', 'purchaseOrder');
+select createDoctype(26, 'PI',   'PI', 'PI', 'Purchase Order Item', 'poitem',  'poitem_id',  'pohead_number',      'vend_name',     'item_number',
+                     '', 'join pohead on poitem_pohead_id=pohead_id join vendinfo on pohead_vend_id=vend_id ' ||
+                     'join itemsite on poitem_itemsite_id=itemsite_id join item on itemsite_item_id=item_id');
+select createDoctype(27, 'RA',   'RA', 'RA', 'Return Authorization', 'rahead',  'rahead_id',  'rahead_number',       'cust_name',     'firstline(rahead_comments)',
```
Comment (bendiy) [inline foundation-database/public/patches/populate_source.sql]: `firstline(rahead_notes)` not `firstline(rahead_comments)`?

File: foundation-database/public/patches/populate_source.sql (hunk 1 above)
Comment (gilmoskowitz) [inline foundation-database/public/patches/populate_source.sql]: Correct - the "destination warehouse" as set in the `tohead` table.

File: foundation-database/public/patches/populate_source.sql [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                     '', 'join item on bomhead_item_id = item_id', 'bomhead_id', 'bom');
+select createDoctype( 3, 'BBI',  'BBI', 'BBI',   'Breeder BOM Item', 'bbomitem', 'bbomitem_id', 'bbomitem_docnum',  'item_number',   'firstline(item_descrip1)',
+                     '', 'join item on bbomitem_item_id = item_id', 'bbomitem_id', 'bbom');
+select createDoctype( 4, 'BMH',  'BMH', 'BMH',   'BOM Head',         'bomhead',  'bomhead_id',  'bomhead_docnum',   'item_number',   'firstline(item_descrip1)',
+                     '', 'join item on bomhead_item_id = item_id', 'bomhead_id', 'bom');
+select createDoctype( 5, 'BMI',  'BMI', 'BMI',   'BOM Item',         'bomitem', 'bomitem_id', 'p.item_number',      'c.item_number', 'firstline(c.item_descrip1)',
+                     '', 'join item p on bomitem_parent_item_id = p.item_id join item c on bomitem_item_id = c.item_id');
+select createDoctype( 6, 'BOH',  'BOH', 'BOH',   'Routing Head',     'booitem', 'booitem_id', 'p.item_number',      'c.item_number', 'firstline(c.item_descrip1)',
+                     '', 'join item p on booitem_parent_item_id = p.item_id join item c on bomitem_item_id = c.item_id');
```
Comment (bendiy) [inline foundation-database/public/patches/populate_source.sql]: `booitem_parent_item_id` doesn't exist

File: foundation-database/public/patches/populate_source.sql [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+select createDoctype( 7, 'BOI',  'BOI', 'BOI',   'Routing Item',     'booitem', 'booitem_id', 'p.item_number',      'c.item_number', 'firstline(c.item_descrip1)',
+                     '', 'join item p on booitem_parent_item_id = p.item_id join item c on booitem_item_id = c.item_id');
```
Comment (bendiy) [inline foundation-database/public/patches/populate_source.sql]: `booitem_parent_item_id` doesn't exist

File: foundation-database/public/patches/populate_source.sql [hunk 5]
```diff
@@ -0,0 +1,96 @@
+-- source col: docass_num name  docass charass descrip key_field
+--                   number_field name_field, desc_field widget
+--                   pJoin pParam pUi pPriv
+-- TATC for Time Attendance?
+select createDoctype( 1, 'ADDR', 'ADDR', 'ADDR',   'Address',          'addr',     'addr_id',     'addr_number',      'addr_line1',    'addr_line2');
+select createDoctype( 2, 'BBH',  'BBH', 'BBH',   'Breeder BOM Head', 'bbomhead', 'bbomhead_id', 'bbomhead_docnum',  'item_number',   'firstline(item_descrip1)',
+                     '', 'join item on bomhead_item_id = item_id', 'bomhead_id', 'bom');
+select createDoctype( 3, 'BBI',  'BBI', 'BBI',   'Breeder BOM Item', 'bbomitem', 'bbomitem_id', 'bbomitem_docnum',  'item_number',   'firstline(item_descrip1)',
```
Comment (bendiy) [inline foundation-database/public/patches/populate_source.sql]: `bbomitem_docnum` doesn't exist

**Pull request merged**